### ISSUE TITLE
RFairLock dead lock issue

### ISSUE THREAD

**Issue by mjavault** [redisson/redisson#1104]
Hi,
I would like to report a locking issue we are seeing using redisson versions 3.3.2 to 3.5.4, using `RFairLock`.
We have 3 instances of an application competing for a fair lock. After some time, the timeout of the fair lock queue gets set to a time many hours in the future, and no thread ever gets the lock.

Here is what redis is reporting during one of the occurences - both commands ran at about 9:30AM PST on 10/17/2017 (`1508257800000`):
```
redis > zrange "redisson_lock_timeout:{mylock}" 0 99 WITHSCORES
1) "3b9c4448-dfaa-4b4e-a691-cfee25458333:88"
2) "1508278998141"
3) "a8baa8c5-d42d-4912-9a1b-e72d0728ae88:90"
4) "1508278998141"
5) "17cd3d41-8ea1-44c7-b458-a530aff209f0:88"
6) "1508279003141"
```
```
redis > lrange "redisson_lock_queue:{mylock}" 0 99
1) "17cd3d41-8ea1-44c7-b458-a530aff209f0:88"
2) "3b9c4448-dfaa-4b4e-a691-cfee25458333:88"
3) "a8baa8c5-d42d-4912-9a1b-e72d0728ae88:90"
```

You can see that the timeout (zscore of the timeout keys) is set about 6 hours in the future.
```Current time = Tuesday, October 17, 2017 9:30:00.000 AM GMT-07:00 DST```
```1508278998141 = Tuesday, October 17, 2017 3:23:18.141 PM GMT-07:00 DST```
```1508279003141 = Tuesday, October 17, 2017 3:23:23.141 PM GMT-07:00 DST```

So far, we haven't been able to reproduce the issue locally, but it is happening daily in our production environment. We are suspecting the problem is happening randomly when one of the application gets killed by the operating system (kill -9), something that can happen quite often due to the way we manage the apps.

We were using redisson version 3.2.3 before, and never experienced the issue.
Earlier this year, we tried updating to version 3.3.2 and started seeing the problem within a day or so. We reverted to 3.2.3, and waited for an updated version.
Two days ago, we decided to upgrade to 3.5.4 assuming the problem was fixed, but it happened again within a few hours.
I will keep on trying to reproduce the issue, but any help is appreciated.
Thanks,
Michael

**Comment by mjavault** [comment]
I managed to reproduce the issue locally. I will try to post a code that does it all the time.
In the mean time, I noticed a pretty big drift in the queue timeouts that I can't explain (redisson 3.5.4):
```
time
pttl test-fair-lock
zrange "redisson_lock_timeout:{test-fair-lock}" 0 99 WITHSCORES
1508367800								4:03:20 PM GMT-07:00 DST
376547
25952									25s
6ddb3f73-59db-41a1-bad7-e6c6c6f633c1:1
1508367851327							4:04:11.327 PM GMT-07:00 DST
6fe889c6-7496-4c53-ad4f-c8850e3a0d2e:1
1508367851327							4:04:11.327 PM GMT-07:00 DST
7cb68435-b1c6-41d4-ba07-fb6eee02c9d7:1
1508367856327							4:04:16.327 PM GMT-07:00 DST
```
```
time
pttl test-fair-lock
zrange "redisson_lock_timeout:{test-fair-lock}" 0 99 WITHSCORES
1508367803								4:03:23 PM GMT-07:00 DST
793817
22535									22s
7cb68435-b1c6-41d4-ba07-fb6eee02c9d7:1
1508367876327							4:04:36.327 PM GMT-07:00 DST
6ddb3f73-59db-41a1-bad7-e6c6c6f633c1:1
1508367881327							4:04:41.327 PM GMT-07:00 DST
6fe889c6-7496-4c53-ad4f-c8850e3a0d2e:1
1508367881327							4:04:41.327 PM GMT-07:00 DST (+55s)
```
```
time
pttl test-fair-lock
zrange "redisson_lock_timeout:{test-fair-lock}" 0 99 WITHSCORES
1508367808								4:03:28 PM GMT-07:00 DST
384414
17944									18s
6ddb3f73-59db-41a1-bad7-e6c6c6f633c1:1
1508367911327							4:05:11.327 PM GMT-07:00 DST
6fe889c6-7496-4c53-ad4f-c8850e3a0d2e:1
1508367911327							4:05:11.327 PM GMT-07:00 DST
7cb68435-b1c6-41d4-ba07-fb6eee02c9d7:1
1508367916327							4:05:16.327 PM GMT-07:00 DST (+90s)
```
```
time
pttl test-fair-lock
zrange "redisson_lock_timeout:{test-fair-lock}" 0 99 WITHSCORES
1508367824								4:03:44 PM GMT-07:00 DST
506066
1822									2s
6ddb3f73-59db-41a1-bad7-e6c6c6f633c1:1
1508368031327							4:07:11.327 PM GMT-07:00 DST
6fe889c6-7496-4c53-ad4f-c8850e3a0d2e:1
1508368031327							4:07:11.327 PM GMT-07:00 DST
b20f0feb-ade7-4d1e-8273-2306ecbef066:1
1508368036327							4:07:16.327 PM GMT-07:00 DST
```

In the scenario above, the lock is taken by another thread that has been killed, so it will expire after the lease time of 30s.
As you can see, over time, the scores of the items in the `redisson_lock_timeout` queue keep on getting larger and larger, which causes an unnecessary delay after the lock has expired if the first thread in the queue has been killed. We now have a placeholder in the queue that will force the others to wait for minutes (eventually hours) before anyone can take the lock.

One thing that looks very suspicious to me is that all the timeouts have the same milliseconds value (.327) over time. I don't see an explanation for that behavior in the code.

**Comment by mjavault** [comment]
Bug identified, introduced with the following commit 1b9ec19cb5ca9e26b39dd84371769761b9906185 #757.
The bug is happening in the following scenario:
- multiple threads are trying to get the same lock, in a loop. This causes the timeout of the thread (score in `redisson_lock_timeout`) to drift quickly away from a normal value.
- the `waitTime` is set to less than 5 seconds.
- the thread at the top of the queue dies, preventing the other elements from the queue to take the lock until it expires - but since the timeout is set to a value far away in the future, this can take a very long time to recover from.

Here is  description of what is happening:
1. Let's assume the lock `L` has been taken by a thread, with a lease of 30 seconds.
2. At T+0, three other threads are now trying to get the same lock, with thread id `A`, `B` and `C` (in that order, let's assume 5ms apart from each other, at T+0, T+5 and T+10 resp), with a try time of 1 second.
When they do so, they get added to `redisson_lock_queue{L}` and to `redisson_lock_timeout{L}` with a score computed as follow:
```
Queue order: A, B, C
Score(A1) = Now() + pttl(L) + 5000
Score(B1) = Score(A1) + 5000
Score(C1) = Score(A1) + 5000
```
3. At T+1000, `A` fails to get the lock and gets removed from both `redisson_lock_queue` and `redisson_lock_timeout` because of the recent commit.
```
Queue order: B, C
Score(B1) = Score(A1) + 5000
Score(C1) = Score(A1) + 5000
```
4. Then `A` tries to take the lock again (before `B` and `C` have expired). The queue and timeout get updated as follow:
```
Queue order: B, C, A
Score(B1) = Score(A1) + 5000
Score(C1) = Score(A1) + 5000
Score(A2) = Score(B1) + 5000 = Score(A1) +5000 + 5000
```
5. At T+1005, `B` expires, gets removed from the queue, and then tries to get the lock again:
```
Queue order: C, A
Score(C1) = Score(A1) + 5000
Score(A2) = Score(B1) + 5000 = Score(A1) +5000 + 5000
```
```
Queue order: C, A. B
Score(C1) = Score(A1) + 5000
Score(A2) = Score(B1) + 5000 = Score(A1) +5000 + 5000
Score(B2) = Score(C1) + 5000 = Score(A1) +5000 + 5000
```
6. At T+1010, `C` expires, gets removed from the queue, and then tries to get the lock again:
```
Queue order: A. B
Score(A2) = Score(B1) + 5000 = Score(A1) +5000 + 5000
Score(B2) = Score(C1) + 5000 = Score(A1) +5000 + 5000
```
```
Queue order: A. B, C
Score(A2) = Score(B1) + 5000 = Score(A1) +5000 + 5000
Score(B2) = Score(C1) + 5000 = Score(A1) +5000 + 5000
Score(C2) = Score(A2) + 5000 = Score(A1) +5000 + 5000
```

Because the order of the elements in `redisson_lock_queue` rapidly changes over time, and is not persistent across retries (i.e. when a thread is actively trying to get a lock, over and over until it does), you end up with the timeout of the thread set to `(number of threads - 1) * 5000` for every `waitTime`. If `waitTime` is small, the timeouts will increase very rapidly.
This by itself is not a problem when all threads never die and remain active. However, if the first thread from the queue dies without cleaning up after itself, the lock will remain blocked until the elements expires, and that expiry will take a very long time to happen because of the drift.
This also shouldn't be a problem for wait time greater than 5000ms, but I would need to confirm this.

**Comment by mrniko** [comment]
Could you please write test for this case?

**Comment by mjavault** [comment]
There you go, full test consistently reproducing the issue.
```
    @Test
    public void testRedissonBug1104() throws Exception {
        Config config = new Config();
        config.useSingleServer().setAddress("redis://127.0.0.1:6379");
        RedissonClient redisson = Redisson.create(config);
        RLock lock = redisson.getFairLock("test-fair-lock");
        AtomicBoolean lastThreadTryingToLock = new AtomicBoolean(false);

        //before we start, make sure we haven't left dead entries from a previous run
        redisson.getScript().eval(RScript.Mode.READ_WRITE, "return redis.call('del', KEYS[1], KEYS[2], KEYS[3])", RScript.ReturnType.INTEGER, Arrays.asList(
                "redisson_lock_timeout:{test-fair-lock}",
                "redisson_lock_queue:{test-fair-lock}",
                "test-fair-lock"
        ));

        //create a scenario where the same 3 threads keep on trying to get a lock
        //to exacerbate the problem, use a very short wait time and a long lease time
        //this will end up setting the queue timeout score to a value far away in the future
        ExecutorService executor = Executors.newFixedThreadPool(3);
        for (int i = 0; i < 100; i++) {
            executor.submit(() -> {
                try {
                    if (lock.tryLock(500, 30000, TimeUnit.MILLISECONDS)) {
                        log.info("Lock taken by thread " + Thread.currentThread().getId());
                        Thread.sleep(30000);
                        try {
                            //this could fail before use sleep for the same value as the lock expiry, that's fine
                            //for the purpose of this test
                            lock.unlock();
                            log.info("Lock released by thread " + Thread.currentThread().getId());
                        } catch (Exception ignored) {
                        }
                    }
                } catch (InterruptedException ex) {
                    log.warn("Interrupted");
                } catch (Exception ex) {
                    log.error(ex, ex);
                }
            });
            //for the first 3 threads, add a 50ms delay. This is to recreate the worst case scenario, where all threads
            //attempting to lock do so in a staggered pattern. This delay will be carried over by the thread pool.
            if (i < 3) {
                Thread.sleep(50);
            }
        }
        //we now launch one more thread and kill it before it manages to fail and clean up
        //that thread will end up with a timeout that will prevent any others from taking the lock for a long time
        executor.submit(() -> {
            log.info("Final thread trying to take the lock with thread id: " + Thread.currentThread().getId());
            try {
                lastThreadTryingToLock.set(true);
                if (lock.tryLock(30000, 30000, TimeUnit.MILLISECONDS)) {
                    log.info("Lock taken by final thread " + Thread.currentThread().getId());
                    Thread.sleep(1000);
                    lock.unlock();
                    log.info("Lock released by final thread " + Thread.currentThread().getId());
                }
            } catch (InterruptedException ex) {
                log.warn("Interrupted");
            } catch (Exception ex) {
                log.error(ex, ex);
            }
        });
        //now we wait for all others threads to stop trying, and only the last thread is running
        while (!lastThreadTryingToLock.get()) {
            Thread.sleep(100);
        }
        //try to kill that last thread, and don't let it clean up after itself
        executor.shutdownNow();
        //force the lock to unlock just in case
        try {
            lock.forceUnlock();
        } catch (Exception ignored) {
        }
        if (lock.isLocked()) {
            fail("Lock should have been unlocked by now");
        }
        //check the timeout scores - they should all be within a reasonable amount of time from now
        List<Long> queue = redisson.getScript().eval(RScript.Mode.READ_ONLY,
                "local result = {}; " +
                        "local timeouts = redis.call('zrange', KEYS[1], 0, 99, 'WITHSCORES'); " +
                        "for i=1,#timeouts,2 do " +
                        "table.insert(result, timeouts[i+1]); " +
                        "end; " +
                        "return result; ",
                RScript.ReturnType.MULTI,
                Collections.singletonList("redisson_lock_timeout:{test-fair-lock}"));
        int i = 0;
        for (Long timeout : queue) {
            log.info("Item " + (i++) + " expires in " + ((timeout - new Date().getTime()) / 1000) + " seconds");
        }

        //and we try one more time to get the lock. At this point, this call should be ideally immediate, but it can take up
        //to 5000ms for the elements in the queue to expire
        log.info("Now trying to get the lock after all threads are done trying, thread id: " + Thread.currentThread().getId());
        if (!lock.tryLock(5000, 1000, TimeUnit.MILLISECONDS)) {
            fail("It took more than 5s to get the lock!");
        }
    }
```

Example of output:
```
2017-10-20 10:54:12.923 INFO  main  [org.redisson.Version] Redisson 3.5.4
2017-10-20 10:54:13.631 INFO  redisson-netty-1-2  [org.redisson.connection.pool.MasterPubSubConnectionPool] 1 connections initialized for /127.0.0.1:6379
2017-10-20 10:54:13.631 INFO  redisson-netty-1-1  [org.redisson.connection.pool.MasterConnectionPool] 10 connections initialized for /127.0.0.1:6379
2017-10-20 10:54:13.716 INFO  pool-3-thread-1  [KeyMutexCacheRedissonTest] Lock taken by thread 33
2017-10-20 10:54:38.430 INFO  pool-3-thread-3  [KeyMutexCacheRedissonTest] Final thread trying to take the lock with thread id: 36
2017-10-20 10:54:38.481 WARN  pool-3-thread-1  [KeyMutexCacheRedissonTest] Interrupted
2017-10-20 10:54:38.481 WARN  pool-3-thread-2  [KeyMutexCacheRedissonTest] Interrupted
2017-10-20 10:54:38.481 WARN  pool-3-thread-3  [KeyMutexCacheRedissonTest] Interrupted
2017-10-20 10:54:38.489 INFO  main  [KeyMutexCacheRedissonTest] Item 0 expires in 500 seconds
2017-10-20 10:54:38.490 INFO  main  [KeyMutexCacheRedissonTest] Item 1 expires in 505 seconds
2017-10-20 10:54:38.490 INFO  main  [KeyMutexCacheRedissonTest] Now trying to get the lock after all threads are done trying, thread id: 1
```

**Comment by mjavault** [comment]
I just tested that code on Redisson version 3.2.3, and the values are slightly higher than I expected:
```
2017-10-20 10:59:54.669 INFO  main  [KeyMutexCacheRedissonTest] Item 0 expires in 10 seconds
2017-10-20 10:59:54.669 INFO  main  [KeyMutexCacheRedissonTest] Item 1 expires in 15 seconds
```
But the results are consistent over time, and they don't exponentially increase. You might want to update the test to target 15s instead of 5s. Actually, based on how you compute the timeout, you probably want to make it `leaseTime + 5s`.

**Comment by mjavault** [comment]
As a temporary fix for us, I have forked and removed the override introduce in 1b9ec19cb5ca9e26b39dd84371769761b9906185:
```
@Override
protected RFuture<Void> acquireFailedAsync(long threadId) {
    (...)
}
```
This fixes the behavior for RedissonFairLock. It consequently reopens #757 but at least we are back in a stable (albeit slow) situation.

**Comment by mrniko** [comment]
Fixed! Thanks for test case and so detailed explanation of the cause.

**Comment by jmchuster** [comment]
Hi all, we bumped our sdk to use `3.7.0` back in May, but saw the same (or similar?) issue again today in production.  Have there been any similar reports or potentially related fixes since 3.7.0?
